Add support for default Ingress Cert

VaishnaviHire opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Jira Issue: https://issues.redhat.com/browse/RHOAIENG-4221 
## Description
<!--- Describe your changes in detail -->

## How Has This Been Tested?

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

zdtsw commented on apis/infrastructure/v1/cert_types.go:
```diff
@@ -17,7 +18,8 @@ type CertificateSpec struct {
-	// +kubebuilder:default=SelfSigned
+	// * DefaultIngress: Default ingress certificate configured for OpenShift
+	// +kubebuilder:validation:Enum=SelfSigned;Provided;DefaultIngress
+	// +kubebuilder:default=DefaultIngress
```
if we change the default value, it can cause problem for these users did not set it in old release.

VaishnaviHire replied:
I think this should be okay. By default users will use default certs. I guess we need to document this change. @israel-hdez Can you confirm?

Jooho replied:
yes  defaultIngress would be the default one for users. If we document it, it would be enough.

VaishnaviHire replied:
Yep, if there is an existing DSCI/DSC, it won't be updated. Currently this has been introduced in dsc

zdtsw commented on pkg/feature/cert.go:
```diff
@@ -126,3 +123,114 @@ func generateCertificate(addr string) ([]byte, []byte, error) {
 	return certBuffer.Bytes(), keyBuffer.Bytes(), nil
 }
+
+// GetDefaultIngressCertificate copies ingress cert secrets from openshift-ingress ns to given namespace.
```
so we have two different ns: openshift-ingress is only to get secret, and openshift-ingress-operator is for ingresscontroller CR?

zdtsw commented on pkg/feature/cert.go:
```diff
@@ -9,24 +9,21 @@ import (
-		return nil
-	}
-
+func (f *Feature) CreateSelfSignedCertificate(secretName string, domain, namespace string) error {
```
```suggestion
func (f *Feature) CreateSelfSignedCertificate(secretName, domain, namespace string) error {
```

zdtsw commented on pkg/feature/serverless/resources.go:
```diff
@@ -1,9 +1,17 @@
 	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/feature"
 )
 
 func ServingCertificateResource(f *feature.Feature) error {
```
is the whole implemenation only applied when Serverless is enabled and only for kserve at the moment?

VaishnaviHire replied:
Yes, only applied to Kserve for now

Jooho commented:
/lgtm

VaishnaviHire commented:
@Jooho @israel-hdez  Is there any changes that we need in odh-model-controller?

bartoszmajsak reviewed (changes requested):
I shared a few minor comments in the code itself, but there's one big thing we should also address - I don't see a reason why new cert/secret logic should be coupled with `Feature` struct. I provided a patch you could apply, as this is certainly way too much to comment (even with suggestions).

PTAL at this [patch](https://gist.github.com/bartoszmajsak/c0f13f86f6a35b2e52754ddf59607938). 

Happy to explain the details. The original code (`CreateSelfSignedCertificate`) was clearly a missed opportunity in the original PR so I felt responsible for cleaning it up :)

Besides that, a couple of tests would really help in gaining confidence that all of it will work as intended. WDYT about adding a few e2e tests covering this as part of this PR?

israel-hdez reviewed (commented):
This almost worked.
The missing part I identify is to set the right secret name in the Knative gateway. As it is, if the user does not specify the SecretName, a default name will be set. The name is set here:

https://github.com/opendatahub-io/opendatahub-operator/blob/df469bbd9a511bb1fb769340b8c5df968fd5c3a5/pkg/feature/serverless/loaders.go#L19

...related template: https://github.com/opendatahub-io/opendatahub-operator/blob/df469bbd9a511bb1fb769340b8c5df968fd5c3a5/components/kserve/resources/servicemesh/routing/istio-ingress-gateway.tmpl.yaml#L17

I can think two ways of doing it:
* Modify that `ServingDefaultValues` so that it will use the name of the copied secret..... Although, looks like `ServingDefaultValues` func runs before the secret is copied, so it is hard to know what's going to be its name. So... IDK.
* When copying the secret, use `f.Spec.KnativeCertificateSecret` contents as the name of the copied secret.

Without that change, models are failing to start on my trials. But once the right secret is configured in the Gateway, it is possible to validate these changes by:

* Installing the operator.
* Installing KServe using the DSC.
* Deploying any sample model.
* Get/download the public certificate that the URL of the deployed model is exposing (hint: https://stackoverflow.com/a/7886248).
* Get/download the public certificate of some other Route (e.g. the OpenShift Console URL).
* Compare both downloaded certificates and check that they are the same.

VaishnaviHire commented on controllers/datasciencecluster/datasciencecluster_controller.go:
```diff
@@ -436,26 +437,17 @@ func (r *DataScienceClusterReconciler) SetupWithManager(mgr ctrl.Manager) error
 		Watches(&source.Kind{Type: &corev1.ConfigMap{}}, handler.EnqueueRequestsFromMapFunc(r.watchDataScienceClusterResources), builder.WithPredicates(configMapPredicates)).
 		Watches(&source.Kind{Type: &apiextensionsv1.CustomResourceDefinition{}}, handler.EnqueueRequestsFromMapFunc(r.watchDataScienceClusterResources),
 			builder.WithPredicates(argoWorkflowCRDPredicates)).
+		Watches(&source.Kind{Type: &corev1.Secret{}}, handler.EnqueueRequestsFromMapFunc(r.watchDefaultIngressSecret), builder.WithPredicates(defaultIngressCertSecretPredicates)).
```
Yep this watch was added to reconcile operator when a certificate is expired and secret gets deleted

israel-hdez replied:
You just need to be careful. We've been seeing OOMKills with clusters with many secrets.

VaishnaviHire commented on pkg/feature/cert.go:
```diff
@@ -126,3 +123,114 @@ func generateCertificate(addr string) ([]byte, []byte, error) {
+		client.InNamespace("openshift-ingress-operator"),
+	}
+
+	err := cli.List(context.TODO(), defaultIngressCtrlList, listOpts...)
```
Yeah I was not sure if there would be an edge case where the `default` is named as something else

VaishnaviHire commented on controllers/datasciencecluster/datasciencecluster_controller.go:
```diff
@@ -465,19 +457,15 @@ func (r *DataScienceClusterReconciler) watchDataScienceClusterForDSCI(a client.O
-	default:
-		return nil
+
+	if a.GetObjectKind().GroupVersionKind().Kind == "CustomResourceDefinition" || a.GetName() == "ArgoWorkflowCRD" {
+		return []reconcile.Request{{
+			NamespacedName: types.NamespacedName{Name: requestName},
+		}}
 	}
```
Yeah I just wanted to reuse code for request name so added a function r.getRequestName()

VaishnaviHire commented:
hi @bartoszmajsak @israel-hdez I have addressed most comments apart from ginkgo tests, getting some issues around those. But this should be good to review in the meantime.

israel-hdez commented on pkg/cluster/cert.go:
```diff
@@ -0,0 +1,226 @@
+		client.InNamespace("openshift-ingress-operator"),
+	}
+
+	err := c.List(ctx, defaultIngressCtrlList, listOpts...)
```
I somehow thought this was going to be changed to a `Get` by name.
Picking the first item in the list will work as long as only the default controller is there. But if users have created their own ingresses, this may no longer pick the OpenShift default one.

VaishnaviHire replied:
If that's the expectation to just get from `default` instance, I will update this.

israel-hdez replied:
Yes, IMO always getting the `default` one should be the expectation.

israel-hdez reviewed (approved):
FYI, I'm getting a `panic` in `GetRelease` func when using `make deploy`.
I had to manually add the CI envvar to prevent the panic.

I tried, and it works. I left a few additional comments.
I anyway approve.

VaishnaviHire commented:
@zdtsw @israel-hdez @bartoszmajsak  I have added e2e tests and also addressed all the comments.

zdtsw commented:
/test opendatahub-operator-e2e

zdtsw commented:
/lgtm
